schema get

ashleygwilliams opened the pull request:
🧣 do no merge 🧣 

fixes #1 

this PR introduces a new crate to the workspace, `rover-client`, which exposes a send function to make graphql requests

ashleygwilliams committed:
workspace: add rover-client

ashleygwilliams committed:
feat(command): add schema

ashleygwilliams committed:
ashley's final push

JakeDawkins committed:
redo module declarations

JakeDawkins committed:
skeleton for schema::get

JakeDawkins committed:
temporarily disable missing_docs lint

JakeDawkins committed:
add default uri to rover-client "Client"

JakeDawkins committed:
clean up WIP "get"

JakeDawkins committed:
add comments explaining graphql-client

JakeDawkins committed:
move client construction to rover

JakeDawkins committed:
WIP

JakeDawkins committed:
fetch command works for sdl

JakeDawkins committed:
fix formatting

JakeDawkins commented on crates/rover-client/src/query/schema/get.rs:
```diff
@@ -0,0 +1,43 @@
+
+    // if we want json, we can parse & serialize it here
+
+    Ok(sdl)
```
I'm not sure what to do about error handling for data that's technically allowed to be optional. Like in the response for the get schema query, we need to access `data.service.schema.document` to get the sdl. But `data`, `service`, and `schema` can all technically be `None`.

Is there a good rusty way for accessing nested data like that other than what I do here? Is it possible to integrate these kinds of errors with `RoverClientError`?

EverlastingBugstopper replied:
im not entirely sure what the original `data` type here looks like so i'd need to spend some time with that, but we likely dont want `expect` here as it will cause this result to panic if it doesn't exist (and it sounds like it's optional?) one thing you can do is call `.ok()` on your `Result` types which will return an `Option`, with `Some(T)` if the `Result` was `Ok` and `None` if it wasn't.

I'm not entirely sure if this solves your problem here since you'll still need to figure out what to do when it returns `None`.

JakeDawkins replied:
Yeah, it's a little hard to explain over text, but the CLI should _fail_ on any of those cases where anything along the way is `None`. I'm not quite sure how to best make it fail in a _nice_ way other than panicking like I did here. This is definitely what I want to pair on in the next couple days sometime

JakeDawkins replied:
I updated all the errors to be a little more explicit, and hopefully this prevents any of the panics :) I'd still like feedback on this, since it's a little verbose, and I'm not sure if there are any patterns to handle this kind of nested data structures well already

JakeDawkins committed:
remove unneeded files

EverlastingBugstopper commented on crates/rover-client/schema.graphql:
```diff
@@ -0,0 +1,7034 @@
+"""
```
should this be in .gitignore? it looks like something you were using to test?

JakeDawkins replied:
I think until we have a strategy for automatically fetching the schema, we should keep this file checked in. This is the file that `graphql-client` uses to generate types for the queries in `rover-client`

EverlastingBugstopper commented on crates/rover-client/src/blocking/client.rs:
```diff
@@ -0,0 +1,65 @@
+impl Client {
+    /// Construct a new [Client] from 2 strings, an `api_key` and an optional `uri`.
+    /// if the uri is None, we'll use the default uri for apollo studio
+    pub fn new(api_key: String, uri: String) -> Client {
```
the comment says that uri is optional here - should this fn take an `Option<String>` so we can see here what the default is? i'm not 100% opinionated either way, default values may make more sense in the CLI side anyway

JakeDawkins replied:
The comment was old. I was originally doing the default logic in the `new` function, but I instead just made the `uri` required and removed the default uri logic for now. I probably want to move the default logic to the session where we initialize everything rather than here, just so if there's any other use case for this `Client`, it wouldn't have the same defaults.

EverlastingBugstopper commented on crates/rover-client/src/error.rs:
```diff
@@ -0,0 +1,30 @@
+#[derive(Error, Debug)]
+pub enum RoverClientError {
+    /// The provided GraphQL was invalid.
+    #[error("encountered a graphQL error, registry responded with: {msg}")]
```
```suggestion
    #[error("encountered a GraphQL error, registry responded with: {msg}")]
```

EverlastingBugstopper commented on crates/rover-client/src/headers.rs:
```diff
@@ -0,0 +1,32 @@
+    let mut headers = reqwest::header::HeaderMap::new();
+
+    let content_type =
+        reqwest::header::HeaderValue::from_str(CONTENT_TYPE).expect("is valid header");
```
might make this a bit readable to do a `use reqwest::header::HeaderValue` up at the top so you can just use `HeaderValue::from_str` directly here

JakeDawkins replied:
Did that for this and for `HeaderMap` :)

EverlastingBugstopper commented on crates/rover-client/src/headers.rs:
```diff
@@ -0,0 +1,32 @@
+        &env::var("CARGO_PKG_VERSION").unwrap_or_else(|_| String::from("unknown")),
+    )
+    .expect("is valid header");
+    headers.insert("apollographql-client-version", client_version);
```
what are these headers required for? is it documented/can we add comments explaining?

EverlastingBugstopper commented on src/cli.rs:
```diff
@@ -13,12 +13,15 @@ pub struct Rover {
     pub fn run(self) -> Result<()> {
         match self.command {
             Command::Config(config) => config.run(),
+            Command::Schema(s) => s.run(),
```
nit
```suggestion
            Command::Schema(schema) => schema.run(),
```

EverlastingBugstopper commented on src/command/schema/mod.rs:
```diff
@@ -0,0 +1,24 @@
+impl Schema {
+    pub fn run(&self) -> Result<()> {
+        match &self.command {
+            Command::Fetch(f) => f.run(),
```
nit
```suggestion
            Command::Fetch(fetch) => fetch.run(),
```

EverlastingBugstopper reviewed (commented):
took a quick look at what's here, i think it would be really helpful for me if we could update the issue and the pr descriptions to describe the proposed usage and the problem that's being solved. i have a very vague idea of what this is trying to solve but would love if we could make it more explicit 😄

JakeDawkins committed:
update with review fixes & better commenting

JakeDawkins committed:
fix outdated client `new` comment;

JakeDawkins committed:
fix formatting

JakeDawkins committed:
made errors explicit, preventing panics

JakeDawkins committed:
fmt

JakeDawkins committed:
wip

JakeDawkins committed:
refactor get fn, add tests

JakeDawkins committed:
lint

JakeDawkins commented on src/command/schema/fetch.rs:
```diff
@@ -0,0 +1,55 @@
+use anyhow::Result;
```
This file is also updated in https://github.com/apollographql/rover/pull/50/files#diff-a3d96ce2322c31552b8ebaf2f6ace8b71024b2070d5eca6a17c99c01904128d9

EverlastingBugstopper replied:
please forgive me if there are comments on this file that we addressed in the other PR - just dismiss them :smile_cat:

EverlastingBugstopper commented on crates/rover-client/Cargo.toml:
```diff
@@ -0,0 +1,15 @@
+[dependencies]
+anyhow = "1.0"
+graphql_client = "0.9"
+http = "0.2.1"
```
```suggestion
http = "0.2"
```

this is a personal thing and we might disagree, but i like to leave out the patch release in sub-crates, that way if there is another version of `http` used in `rover`, we can specify the patch release there and we won't bring in the bloat of two different versions of the same code

EverlastingBugstopper commented on crates/rover-client/Cargo.toml:
```diff
@@ -0,0 +1,15 @@
+anyhow = "1.0"
+graphql_client = "0.9"
+http = "0.2.1"
+reqwest = { version = "0.10.7", features = ["json", "blocking"] }
```
```suggestion
reqwest = { version = "0.10", features = ["json", "blocking"] }
```

EverlastingBugstopper commented on crates/rover-client/Cargo.toml:
```diff
@@ -0,0 +1,15 @@
+http = "0.2.1"
+reqwest = { version = "0.10.7", features = ["json", "blocking"] }
+thiserror = "1.0"
+serde = "1.0.116"
```
```suggestion
serde = "1"
```

serde i go even further and leave out the minor release since i really trust them to not deliver breaking changes. not a hill i'll die on though

EverlastingBugstopper commented on crates/rover-client/Cargo.toml:
```diff
@@ -0,0 +1,15 @@
+reqwest = { version = "0.10.7", features = ["json", "blocking"] }
+thiserror = "1.0"
+serde = "1.0.116"
+serde_json = "1.0.57"
```
```suggestion
serde_json = "1"

```

EverlastingBugstopper commented on crates/rover-client/schema.graphql:
```diff
@@ -0,0 +1,7034 @@
+"""
+Direct the client to resolve this field locally, either from the cache or local resolvers.
+"""
```
is there a source of truth for this file / was it downloaded from somewhere or the result of some nice and fancy codegen? there's just _a lot_ going on here and I'd love to understand how it got here. it's also legit 7k lines... do we need all of it? can we maybe hide our graphql files in a `schemas` directory so i can pretend they don't exist lol

JakeDawkins replied:
This was downloaded from studio, and we could definitely automate the download of this and _not_ commit it, but since that doesn't exist right now, we still need this file _somewhere_. I don't think this is something we should ever have to actually look into. At worst, we should periodically download a new one (maybe using the `fetch` command ;)). At best, we should automate that download and maybe `.gitignore` it.

But I don't have any reservations for the file just existing. It should be the only schema file we ever need

EverlastingBugstopper replied:
perfect!!!!

EverlastingBugstopper commented on crates/rover-client/src/blocking/client.rs:
```diff
@@ -0,0 +1,65 @@
+impl Client {
+    /// Construct a new [Client] from 2 strings, an `api_key` and a `uri`.
+    /// For use in Rover, the `uri` is usually going to be to Apollo Studio
+    pub fn new(api_key: String, uri: String) -> Client {
```
```suggestion
    pub fn new(api_key: &str, uri: &str) -> Client {
```

taking a string slice here will make our API a bit more flexible

![image](https://user-images.githubusercontent.com/9408157/96897480-41485580-1454-11eb-82fd-c08dff3998e6.png)
src https://doc.rust-lang.org/book/ch04-03-slices.html

JakeDawkins replied:
fixed in #50

EverlastingBugstopper commented on crates/rover-client/src/error.rs:
```diff
@@ -0,0 +1,30 @@
+use thiserror::Error;
+
+/// An enum represent all type of possible failures.
```
```suggestion
/// RoverClientError represents all possible failures that can occur during a client request.
```

EverlastingBugstopper commented on crates/rover-client/src/headers.rs:
```diff
@@ -0,0 +1,37 @@
+pub fn build(api_key: &str) -> Result<HeaderMap, RoverClientError> {
+    let mut headers = HeaderMap::new();
+
+    let content_type = HeaderValue::from_str(CONTENT_TYPE).expect("is valid header");
```
```suggestion
    let content_type = HeaderValue::from_str(CONTENT_TYPE)?;
```

this works because of the change i suggested to to the `RoverClientError::InvalidHeaderError` type

EverlastingBugstopper commented on crates/rover-client/src/headers.rs:
```diff
@@ -0,0 +1,37 @@
+    // the metrics in Studio can help us keep track of what parts of the schema
+    // Rover uses and make sure we don't accidentally break those :)
+    // more here: https://www.apollographql.com/docs/studio/client-awareness/#using-apollo-server-and-apollo-client
+    let client_name = HeaderValue::from_str(CLIENT_NAME).expect("is valid header");
```
```suggestion
    let client_name = HeaderValue::from_str(CLIENT_NAME)?;
```

EverlastingBugstopper commented on crates/rover-client/src/headers.rs:
```diff
@@ -0,0 +1,37 @@
+    let client_version = HeaderValue::from_str(
+        &env::var("CARGO_PKG_VERSION").unwrap_or_else(|_| String::from("unknown")),
+    )
+    .expect("is valid header");
```
```suggestion
    )?;
```

EverlastingBugstopper commented on crates/rover-client/src/error.rs:
```diff
@@ -0,0 +1,30 @@
+    HeadersError {
+        /// The error message.
+        msg: http::header::InvalidHeaderValue,
+    },
```
```suggestion
    #[error("invalid header value")]
    InvalidHeader {#[from] http::header::InvalidHeaderValue},
```


this will automatically coerce any `InvalidHeaderValue` into our custom error type and you don't have to worry about passing the message manually

EverlastingBugstopper commented on crates/rover-client/src/error.rs:
```diff
@@ -0,0 +1,30 @@
+    RequestError {
+        /// The error message from [reqwest].
+        msg: reqwest::Error,
+    },
```
```suggestion
    #[error("encountered an error while sending a request}")]
    SendRequest(#[from] reqwest::Error),
```

EverlastingBugstopper commented on crates/rover-client/src/error.rs:
```diff
@@ -0,0 +1,30 @@
+pub enum RoverClientError {
+    /// The provided GraphQL was invalid.
+    #[error("encountered a GraphQL error, registry responded with: {msg}")]
+    GraphQLError {
```
```suggestion
    GraphQL {
```

generally safe to leave `Error` out of these names since that's implied

EverlastingBugstopper replied:
if you use an invalid API token, you get this message, which is not super great `encountered a GraphQL error, registry responded with: 406: Not Acceptable`

should figure out how we can make it inform you that your API key is the problem

EverlastingBugstopper commented on crates/rover-client/src/error.rs:
```diff
@@ -0,0 +1,30 @@
+    },
+    /// Encountered an error handling the received response.
+    #[error("encountered an error handling the response: {msg}")]
+    ResponseError {
```
```suggestion
    HandleResponse {
```

EverlastingBugstopper replied:
so if you type in the wrong graph name, you get this error, which would be nice to clean up a bit:

`encountered an error handling the response: No service found`

if it's the wrong name, the error message could be like `invalid graph name, please enter a valid graph title, which you can find at {insert link}`

EverlastingBugstopper commented on crates/rover-client/src/headers.rs:
```diff
@@ -0,0 +1,37 @@
+    headers.insert("apollographql-client-version", client_version);
+
+    let mut api_key =
+        HeaderValue::from_str(api_key).map_err(|e| RoverClientError::HeadersError { msg: e })?;
```
```suggestion
        HeaderValue::from_str(api_key)?;
```

EverlastingBugstopper commented on crates/rover-client/src/lib.rs:
```diff
@@ -0,0 +1,15 @@
+// #![deny(missing_docs)]
+
+//! HTTP client for making GraphQL requests for the rover CLI tool.
```
```suggestion
//! HTTP client for making GraphQL requests for the Rover CLI tool.
```

EverlastingBugstopper commented on crates/rover-client/src/query/schema/get.rs:
```diff
@@ -0,0 +1,108 @@
+use crate::RoverClientError;
+use graphql_client::*;
+
+// I'm not sure where this should live long-term
+/// this is because of the custom GraphQLDocument scalar in the schema
```
probably wouldn't hurt to just put this at the root of the module. do any of our functions return this type? if so, we should definitely expose it as part of our public API so Rover can consume it properly. if not you can just put it at the top level and `pub(crate)` it

JakeDawkins replied:
Right now, I think this is the only place where this type will be used. If we add another query that fetches a schema document, that could change. Do you think it's fine to leave here until it's actually needed multiple places?

EverlastingBugstopper replied:
yeah! totally fine

EverlastingBugstopper commented on src/command/schema/fetch.rs:
```diff
@@ -0,0 +1,55 @@
+
+                match schema {
+                    Ok(schema) => log::info!("{}", schema),
+                    Err(err) => log::error!("{}", err),
```
should this be a hard error or a warning? perhaps `log::warn` is a better choice here if it does not exit with a non-zero exit code

EverlastingBugstopper commented on src/command/schema/fetch.rs:
```diff
@@ -0,0 +1,55 @@
+    #[structopt(long, default_value = "current")]
+    variant: String,
+    #[structopt(long, default_value = "default")]
+    profile: String,
```
```suggestion
    #[structopt(long = "profile", default_value = "default")]
    profile_name: String,
```

this is consistent with the profile argument we use in `src/command/config/api_key`.

i wonder if there is a way to store these commands with `StructOpt` so it's easier to keep them consistent

EverlastingBugstopper commented on src/command/schema/fetch.rs:
```diff
@@ -0,0 +1,55 @@
+pub struct Fetch {
+    /// ID of the graph to fetch from Apollo Studio
+    #[structopt(name = "SCHEMA_ID")]
+    schema_id: String,
```
this was kind of confusing to me as a user, since what we're really looking for here is the name of the graph in apollo studio. it is technically an ID, but in the UI it is titled `Graph Title`, not `Schema ID`. Might be worth updating this to reflect that verbiage

JakeDawkins replied:
Changing it to `graph_name` instead. I don't think `Title` makes much sense, and when creating a new graph in studio, it asks for a `graph name` instead. I know, the naming is a bit confusing there

EverlastingBugstopper reviewed (changes requested):
looking good! few idiomatic things here and there :smile: let me know if you have questions/need further clarification!

would probably also be good to get a first pass at the docs in here

JakeDawkins committed:
Update crates/rover-client/Cargo.toml

Co-authored-by: Avery Harnish <avery@apollographql.com>

JakeDawkins committed:
Update crates/rover-client/Cargo.toml

Co-authored-by: Avery Harnish <avery@apollographql.com>

JakeDawkins committed:
Update crates/rover-client/Cargo.toml

Co-authored-by: Avery Harnish <avery@apollographql.com>

JakeDawkins committed:
Update crates/rover-client/src/error.rs

Co-authored-by: Avery Harnish <avery@apollographql.com>

JakeDawkins committed:
address review comments

- return from command with error instead of printing to log::error
- fix graph_name naming
- fix profile_name for consistency

JakeDawkins committed:
update error names and usages

JakeDawkins committed:
merge main

EverlastingBugstopper reviewed (approved):
lgtm! also opened an issue for docs at #61
